Documentation for debugging CRD-based Zenith (#145)
stackhpc/azimuth-config: docs/debugging/zenith-services.md

kubectl -n azimuth rollout restart deployment/zenith-server-sshd
```

## Client not appearing in Zenith CRDs

> The components of Zenith communicate using three Kubernetes CRDs:

Reviewer: ```suggestion
The components of Zenith communicate using three [Kubernetes CRDs](https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/):
```

  * `services.zenith.stackhpc.com`  

>     Respresents a reserved domain and stores the associated public key.

Reviewer: ```suggestion
    A reserved domain and associated SSH public key.
```

  * `endpoints.zenith.stackhpc.com`  

>     Stores the current endpoints for a Zenith service.  

Reviewer: ```suggestion
    The current endpoints for a Zenith service.  
```

>     Each SSH tunnel for a service updates this resource to add its address, port and configuration.

Reviewer: Does the SSH tunnel update this? Could be clearer in the mechanism I think.

Author: The script that runs at the server end of the tunnel does indeed update this.

>     Each SSH tunnel for a service updates this resource to add its address, port and configuration.

Reviewer: ```suggestion
    Each Zenith SSH tunnel for a service updates this resource to add its address, port and configuration.
```

>     Each Zenith SSH tunnel for a service updates this resource to add its address, port and configuration.

Reviewer: ```suggestion
    This resource is updated to add the address, port and configuration of the Zenith SSH tunnel as the SSH tunnel is created.
```

  * `leases.zenith.stackhpc.com`  

>     Stores heartbeat information for an individual SSH tunnel.  

Reviewer: ```suggestion
    Heartbeat information for an individual SSH tunnel.  
```

>     Each SSH tunnel has its own lease resource that is regularly updated with a heartbeat.

Reviewer: ```suggestion
    Each Zenith SSH tunnel has its own lease resource that is regularly updated with a heartbeat.
```

> If a service is not functioning as expected, we should check the state of the CRDs for

Reviewer: ```suggestion
If a Zenith service is not functioning as expected, check the state of the CRDs for 
```

$ kubectl -n zenith-services get leases.zenith
NAME                                         RENEWED   TTL   REAP AFTER   AGE
sh20tp1071hl3xtjw5cj4mwdy5t0v7qodj31-fn75d   7s        20    120          13m
igxvo2okpkq834d1qbgtlhmm6xo4laj0dupn-7tnqm   5s        20    120          14m
```

Finally, check that the endpoint is registered correctly in the endpoints resource:

```command  title="On the K3s node, targetting the HA cluster if deployed"
$ kubectl -n zenith-services get endpoints.zenith igxvo2okpkq834d1qbgtlhmm6xo4laj0dupn -o yaml
apiVersion: zenith.stackhpc.com/v1alpha1
kind: Endpoints
metadata:
  creationTimestamp: "2024-05-01T13:24:12Z"
  generation: 3
  name: igxvo2okpkq834d1qbgtlhmm6xo4laj0dupn
  namespace: zenith-services
  ownerReferences:
  - apiVersion: zenith.stackhpc.com/v1alpha1
    blockOwnerDeletion: true
    kind: Service
    name: igxvo2okpkq834d1qbgtlhmm6xo4laj0dupn
    uid: 378b39dc-9fce-4553-865e-edad2dd8d8b0
  resourceVersion: "7260"
  uid: 62badd6e-a5a7-452d-b346-04c2efd75a6c
spec:
  endpoints:
    7tnqm:
      address: 10.42.0.71
      config:
        backend-protocol: http
        skip-auth: false
      port: 42109
      status: passing
```

> The address should be the pod IP of an SSHD pod and the port should be the allocated port

Reviewer: ```suggestion
The address should be the pod IP of a Zenith SSHD pod, and the port should be the allocated port
```

kubectl -n azimuth rollout restart deployment/azimuth-identity-operator
```

If this doesn't work, check the logs for errors:

```sh  title="On the K3s node, targetting the HA cluster if deployed"
kubectl -n azimuth logs deployment/azimuth-identity-operator [-f]
```

## Kubernetes resources for the Zenith service have not been created

If the CRDs for the service look correct, it is possible that the component that watches

> the Zenith CRDs and hooks up the Kubernetes ingress for those services is not functioning

Reviewer: ```suggestion
the Zenith CRDs and creates the Kubernetes ingress for those services is not functioning
```
